# International Breweries Analysis

### Introduction
International Breweries Plc manufactures beer and non-alcoholic malt drinks. Its activities include the brewing, packaging, and marketing of beer, alcoholic flavoured and non-flavoured alcoholic beverages, and soft drinks. The company was founded by Lawrence Omole on December 22, 1971 and is headquartered in Lagos, Nigeria.

### Project Aim:
This project was created for learning purpose. 

* It was design for student, to help them understand how to connect to a database on a local host using the necessary connection strings.
* In the process, they get to familiarize themselves with some SQL queries by using it to query the database for insights.
* Lastly, it shows the student how to visualize the output with Python Plotly interactive library

### Data Description

|Columns|Type|
|---|---|
|Sales_ID|char(5)|	
|Region| varchar(25)|	
|Months| varchar(25)|	
|Year| smallint |	
|Sales_Rep| varchar(25)|	
|Brands| varchar(25)|	
|Unit_Cost| smallint|	
|Unit_Price|smallint|	
|Quantity|smallint|	
|Country| varchar(25)|



#### Connecting to a PostgreSql database on the local host

In [1]:
from sqlalchemy import create_engine
engine = create_engine('postgresql://postgres:28_March@localhost/international_breweries')
%reload_ext sql
%sql $engine.url

##### Drop Columns if the exist already

In [2]:
%%sql 
ALTER TABLE breweries_sales 
DROP COLUMN total_cost,
DROP COLUMN total_sale,
DROP COLUMN profit,
DROP COLUMN roi,
DROP COLUMN shortMnthName;

 * postgresql://postgres:***@localhost/international_breweries
Done.


[]

##### Creating new columns

In [3]:
%%sql
ALTER TABLE breweries_sales
ADD COLUMN total_cost bigint,
ADD COLUMN total_sale bigint,
ADD COLUMN profit bigint,
ADD COLUMN roi NUMERIC(10, 5),
ADD COLUMN shortMnthName character(3);

 * postgresql://postgres:***@localhost/international_breweries
Done.


[]

##### Populating new columns with values

In [4]:
%%sql 
UPDATE breweries_sales 
SET total_cost = unit_cost * CAST(quantity_sold AS INTEGER),
    total_sale = unit_price * CAST(quantity_sold AS INTEGER),
    profit = (unit_price * CAST(quantity_sold AS INTEGER)) - (unit_cost * CAST(quantity_sold AS INTEGER)),
    roi = (((unit_price * CAST(quantity_sold AS INTEGER)) - (unit_cost * CAST(quantity_sold AS INTEGER)))/(unit_cost * CAST(quantity_sold AS INTEGER))::NUMERIC)*100,
    shortMnthName = LEFT(month, 3);

 * postgresql://postgres:***@localhost/international_breweries
1047 rows affected.


[]

#### Preview Table

In [5]:
%%sql
-- Preview
SELECT *
FROM breweries_sales
LIMIT 3

 * postgresql://postgres:***@localhost/international_breweries
3 rows affected.


[('10962', 'northwest', 'October', 2019, 'Jardine', 'trophy', 150, 200, 955, 'Nigeria', 143250, 191000, 47750, Decimal('33.33333'), 'Oct'),
 ('10967', 'southsouth', 'March', 2019, 'Parent', 'beta malt', 80, 150, 797, 'Nigeria', 63760, 119550, 55790, Decimal('87.50000'), 'Mar'),
 ('10969', 'northeast ', 'May', 2019, 'Smith', 'trophy', 150, 200, 773, 'Benin', 115950, 154600, 38650, Decimal('33.33333'), 'May')]

#### Checking Distinct Categorical Columns

In [6]:
%%sql

--Unique Countries

SELECT DISTINCT(country) unique_countries
FROM breweries_sales

 * postgresql://postgres:***@localhost/international_breweries
5 rows affected.


[('Nigeria',), ('Benin',), ('Ghana',), ('Senegal',), ('Togo',)]

In [7]:
%%sql

--Distinct Region

SELECT DISTINCT(region) unique_regions
FROM breweries_sales

 * postgresql://postgres:***@localhost/international_breweries
6 rows affected.


[('southsouth',),
 ('west',),
 ('northcentral ',),
 ('Southeast',),
 ('northeast ',),
 ('northwest',)]

In [8]:
%%sql

--Distinct Sales Representative

SELECT DISTINCT(sales_rep) unique_sales_rep
FROM breweries_sales

 * postgresql://postgres:***@localhost/international_breweries
11 rows affected.


[('Morgan',),
 ('Gill',),
 ('Jones',),
 ('Sorvino',),
 ('Thompson',),
 ('Kivell',),
 ('Jardine',),
 ('Andrews',),
 ('Smith',),
 ('Howard',),
 ('Parent',)]

In [9]:
%%sql

--Unique Brand

SELECT DISTINCT(brand) unique_brand
FROM breweries_sales

 * postgresql://postgres:***@localhost/international_breweries
7 rows affected.


[('beta malt',),
 ('eagle lager',),
 ('grand malt',),
 ('hero',),
 ('budweiser',),
 ('castle lite',),
 ('trophy',)]

#### Importing plotly library for visualization

In [10]:
import pandas as pd
import plotly.express as px
import plotly.offline as po
import plotly.graph_objects as go
import plotly.io as pio
pio.templates.default= 'plotly_dark'
po.init_notebook_mode(connected=True)

# Plotly Themes
# ["plotly", "plotly_white", "plotly_dark", "ggplot2", "seaborn", "simple_white", "none"]

### Yearly Overview

#### Query the database

In [11]:
query = '''
SELECT 
    year, 
    SUM(quantity_sold) total_qty,
    SUM(total_cost) total_cost,
    SUM(total_sale) total_sale,
    SUM(profit) total_profit,
    AVG(roi) roi

FROM 
    breweries_sales
GROUP BY
    year
'''
year_sum = pd.read_sql_query(query, engine)
year_sum

year  total_qty  total_cost  total_sale  total_profit        roi
0  2018     317563  48577780.0  87081100.0    38503320.0  75.632404
1  2019     305409  47422350.0  84486200.0    37063850.0  74.662308
2  2020     267750  40305100.0  70325350.0    30020250.0  71.346237

#### Visualize the result

In [12]:
year_sum.year = year_sum.year.astype('str')

fig = px.bar(year_sum, 'year', ['total_cost', 'total_sale', 'total_profit'], width=700, 
             barmode='group', text_auto='$.2s', labels={'year':'', 'value':''},
             title='Yearly Overview', color_discrete_sequence=px.colors.qualitative.Set1)
fig.update_traces(textposition='outside', cliponaxis=False)
fig.update_yaxes(showticklabels=False)
fig.update_layout(legend = dict(
                        orientation='h',
                        title='',
                        font = dict(
                        family="Courier",
                        size=12,
                        ),
                        bgcolor='olive',
                        bordercolor='blue',
                        borderwidth=.5
                    ))

>**Notes**: 
>
>We can see that there is decrease in across total cost, total sales, and total profit. In 2019, the total sale decrease by 3.4% compared to year 2018. The decrease was a bit much for year 2020, 19.5% compared to year 2018, and 16.67% compared to year 2019 

### Monthly Overview

#### Query the database

In [13]:
query = '''
SELECT 
    shortmnthname, 
    SUM(quantity_sold) total_qty,
    SUM(total_cost) total_cost,
    SUM(total_sale) total_sale,
    SUM(profit) total_profit,
    AVG(roi) roi

FROM 
    breweries_sales
GROUP BY
    shortmnthname
'''
mnth_sum = pd.read_sql_query(query, engine)
mnth_sum

shortmnthname  total_qty  total_cost  total_sale  total_profit        roi
0            Aug      74908  11363860.0  20094800.0     8730940.0  73.363195
1            May      74089  11356250.0  20128500.0     8772250.0  73.923821
2            Nov      74120  11386870.0  20207200.0     8820330.0  73.890015
3            Dec      73538  11298510.0  20085800.0     8787290.0  73.923821
4            Feb      74847  11503400.0  20531850.0     9028450.0  74.598930
5            Jun      73637  11324060.0  20152500.0     8828440.0  74.129478
6            Jul      74591  11320340.0  20107350.0     8787010.0  74.306963
7            Oct      73951  11254550.0  20021650.0     8767100.0  74.546427
8            Jan      74373  11406060.0  20129050.0     8722990.0  73.665886
9            Apr      72562  11103620.0  19677450.0     8573830.0  73.890015
10           Mar      76206  11680670.0  20746800.0     9066130.0  74.234068
11           Sep      73900  11307040.0  20009700.0     8702660.0  73.732251

#### Visualize the result

In [14]:
mnth_sum['shortmnthname'] = pd.to_datetime(mnth_sum['shortmnthname'], format="%b").dt.month
mnth_sum.sort_values('shortmnthname', inplace=True)
mnth_order = ['Jan', 'Feb', 'Mar', 'Apr', 'May', 'Jun', 
              'Jul', 'Aug', 'Sep', 'Oct', 'Nov', 'Dec']

In [15]:

fig = go.Figure()
fig.add_trace(
    go.Scatter(x=mnth_sum.shortmnthname, y=mnth_sum.total_cost, 
               name='Total Cost')
),
fig.add_trace(
    go.Scatter(x=mnth_sum.shortmnthname, y=mnth_sum.total_sale, 
               name='Total Sale')
),
fig.add_trace(
    go.Scatter(x=mnth_sum.shortmnthname, y=mnth_sum.total_profit,
               name='Total Profit')
)
fig.update_layout(title='Monthly Overview', width=700,
                  xaxis=dict(ticktext=mnth_order,
                             tickvals=[1,2,3,4,5,6,7,8,9,10,11,12]),
                  yaxis=dict(tickformat='$.2s'),
                  legend = dict(
                        orientation='v',
                        title='',
                        font = dict(
                        family="Courier",
                        size=12,
                        ),
                        bgcolor='olive',
                        bordercolor='blue',
                        borderwidth=.5
                    )
                  )

>**Notes**: 
>
>COnsidering the whole data, we can see that the total profit across all month is the range of $8M to $10M.

## Brand Analysis
### Distribution of Quantity Sold across Brands

#### Query the database

In [16]:
query = '''
SELECT 
    brand,
    year, 
    SUM(quantity_sold) qty_sold
FROM
    breweries_sales
GROUP BY
    brand,
    year
'''
brand_qty_sld = pd.read_sql_query(query, engine).sort_values(by=['year','qty_sold'])
brand_qty_sld.head()

brand  year  qty_sold
17       trophy  2018     39819
11         hero  2018     41314
0    grand malt  2018     45902
6     budweiser  2018     46720
2   castle lite  2018     47307

#### Visualize the result

In [17]:
fig = px.bar(brand_qty_sld, 'qty_sold', 'brand', facet_col='year', 
             color='qty_sold', title='Quantity Sold by Brand',
             labels={'brand':'', 'qty_sold':'Qty. Sold'}, width=800, height=600,
             text_auto='$.3s', color_continuous_scale='Blues')
fig.update_xaxes(showticklabels=False)

>**Notes**: 
>
> As of 2019, Grand Malt was the least sold brand amongst all brand, a total of 35,202. In the recent year 2020, it is the highest sold brand, a total of 45,185 (22% increase)
>
> Beta Malt which is the least sold brand in 2020 (33,755), was the second brand with the highest quantity sold in year 2018 (48,042), 29.7% decrease.

### Total Cost and Total  Sale by Brand

#### Query the database

In [18]:
query = '''
SELECT 
    brand,
    year, 
    SUM(total_cost) t_cost,
    SUM(total_sale) t_sale
FROM
    breweries_sales
GROUP BY
    brand,
    year
'''
brand_cst_sale = pd.read_sql_query(query, engine).sort_values(by=['year', 't_cost'])
brand_cst_sale.head()

brand  year     t_cost      t_sale
9     beta malt  2018  3843360.0   7206300.0
0    grand malt  2018  4131180.0   6885300.0
17       trophy  2018  5972850.0   7963800.0
11         hero  2018  6197100.0   8262800.0
12  eagle lager  2018  8238030.0  12114750.0

#### Visualize the result

In [19]:
fig = px.bar(brand_cst_sale, ['t_cost', 't_sale'], 'brand', facet_col='year', 
             text_auto='$.3s', title='Total Cost and Total Sale by Brand', height=500, width=700,
             labels={'brand':'', 'value':''},barmode='group',
             color_discrete_sequence=px.colors.qualitative.Set1,
             )
fig.update_traces(textposition='auto')
fig.update_layout(font=dict(size=8))
fig.update_layout(legend = dict(
                        orientation='h',
                        title='',
                        font = dict(
                        family="Courier",
                        size=12,
                        ),
                        bgcolor='olive',
                        bordercolor='blue',
                        borderwidth=.5
                    ))

>**Notes**: 
>
> Budweiser and Castle made larger sales across the three year, compared to other brands. Grand Malt is the only brand that has not made sales above $7M 

### Profit by Brand

#### Query the database

In [20]:
query = '''
SELECT 
    brand,
    year, 
    SUM(profit) profit
FROM
    breweries_sales
GROUP BY
    brand,
    year
'''
brand_profit = pd.read_sql_query(query, engine).sort_values(by=['year','profit'])
brand_profit.head()

brand  year     profit
17       trophy  2018  1990950.0
11         hero  2018  2065700.0
0    grand malt  2018  2754120.0
9     beta malt  2018  3362940.0
12  eagle lager  2018  3876720.0

#### Visualize the result

In [21]:
fig = px.bar(brand_profit, 'profit', 'brand', facet_col='year', text_auto='$.3s',
             color='profit', labels={'brand':'', 'profit':'Profit'}, height=500,
             title='Profit by Brand', color_continuous_scale='blues', width=700)
fig

>**Notes**: 
>
> We can see from the bar charts that **Castle Lite** and **Budweiser** are the two brand that generates the most profit across the three year. The other brands has not made a profit up to half of the aforementioned brand across the three years.
>
> Let's look at the proportion of this profit.

### Proportion of Profit 

#### Query the database

In [22]:
query = '''
WITH proportion_18 AS (
    SELECT
        brand,
        ROUND((SUM(profit)/(
            SELECT
                SUM(profit)
            FROM 
                breweries_sales
            WHERE
                year = 2018
        )) *100, 1) percent_18
    FROM 
        breweries_sales
    WHERE
        year = 2018
    GROUP BY 
        brand
    ORDER BY 
        percent_18
),

proportion_19 AS (
    SELECT
        brand,
        ROUND((SUM(profit)/(
            SELECT
                SUM(profit)
            FROM 
                breweries_sales
            WHERE
                year = 2019
        )) *100, 1) percent_19
    FROM 
        breweries_sales
    WHERE
        year = 2019
    GROUP BY 
        brand
    ORDER BY 
        percent_19
),

proportion_20 AS (
    SELECT
        brand,
        ROUND((SUM(profit)/(
            SELECT
                SUM(profit)
            FROM 
                breweries_sales
            WHERE
                year = 2020
        )) *100, 1) percent_20
    FROM 
        breweries_sales
    WHERE
        year = 2020
    GROUP BY 
        brand
    ORDER BY 
        percent_20
)
    
SELECT 
    brand,
    percent_18, percent_19, percent_20 
FROM
    proportion_18
JOIN
    proportion_19 USING(brand)
JOIN
    proportion_20 USING(brand)
'''

prop = pd.read_sql_query(query, engine).set_index('brand')
prop.index.name = None
prop

percent_18  percent_19  percent_20
beta malt           8.7         8.2         7.9
budweiser          30.3        30.7        28.3
castle lite        33.2        34.0        30.7
eagle lager        10.1         8.6        10.5
grand malt          7.2         5.7         9.0
hero                5.4         6.1         7.1
trophy              5.2         6.7         6.5

#### Visualize the result

In [23]:
fig = px.imshow(prop, width=500, aspect='auto', text_auto=".1f",
                title='Profit proportion for each year', 
                color_continuous_scale=px.colors.sequential.Blues)
fig.update_xaxes(side='top')

>**Notes**: 
>
> We can see that **Budweiser** and **Castle Lite** makes a profit of above 25% in the three years compared to others. For other brand, their profits is below 12% across the three years.
>
> Now let's look at the Average ROI for each brand, we expect Budweiser and Castle Lite to have the highest.

### Average Return of Interest (ROI) by Brand

#### Query the database

In [24]:
query = '''
SELECT 
    brand,
    AVG(roi) avg_roi
FROM 
    breweries_sales
GROUP BY
    brand
'''

brand_roi = pd.read_sql_query(query, engine).sort_values('avg_roi')
brand_roi

brand    avg_roi
3         hero   33.33333
6       trophy   33.33333
1  eagle lager   47.05882
2   grand malt   66.66667
0    beta malt   87.50000
4    budweiser  100.00000
5  castle lite  150.00000

#### Visualize the result

In [25]:
fig =  px.bar(brand_roi, 'avg_roi', 'brand', text_auto='.3s', color='avg_roi',
              color_continuous_scale='blues', labels={'brand':'','avg_roi': 'Avg. ROI(%)'},
              title='Average ROI by Brand', height=500, width=700)
fig.update_xaxes(showticklabels=False)

>**Notes**: 
>
> As expected, **Castle Lite** and **Budweiser** has the highest ROI, 150% and 100% respectively. **Hero** and **Trophy** has the same ROI which happens to be the lowest amongst the brands.

## Country Analysis
### Distribution of Quantity Sold by Country
#### Query the database

In [26]:
query = '''
SELECT 
    country, 
    SUM(quantity_sold) total_qty
FROM 
    breweries_sales
GROUP BY 
    country
ORDER BY 
    total_qty DESC
'''
cntry_total_qty = pd.read_sql_query(query, engine)
cntry_total_qty

country  total_qty
0  Senegal     180928
1  Nigeria     179307
2    Ghana     178255
3    Benin     176324
4     Togo     175908

#### Visualize the result

In [27]:
fig = px.pie(cntry_total_qty, names='country', values='total_qty', width=500,
       color_discrete_sequence=px.colors.qualitative.Set1,
       title='Quantity Sold by Country')
fig.update_traces(textinfo = 'percent+label', showlegend=False, textposition='outside')

>**Notes**: 
>
> The distribution of quantity sold by each country is almost equal. The difference is not above 2%

### Total Cost and Total Sales by Country

#### Query the database

In [28]:
query = '''
SELECT 
    country, 
    year,
    SUM(total_cost) t_cost,
    SUM(total_sale) t_sale
FROM 
    breweries_sales
GROUP BY 
    country,
    year
ORDER BY 
    year DESC, t_cost
'''
cntry_profit = pd.read_sql_query(query, engine)
cntry_profit.head()

country  year     t_cost      t_sale
0    Benin  2020  6927960.0  12201300.0
1  Nigeria  2020  7126630.0  11931950.0
2     Togo  2020  7777190.0  13887150.0
3  Senegal  2020  8792640.0  15480200.0
4    Ghana  2020  9680680.0  16824750.0

#### Visualize the result

In [29]:
fig = px.bar(cntry_profit, 'country', ['t_cost', 't_sale'], text_auto='$.3s', facet_row='year',
             title='Total Cost and Total Sales by Country', height=700, barmode='group',width=800,
             color_discrete_sequence=px.colors.qualitative.Set1, labels={'value':'', 'country':''})
fig.update_traces(textposition='outside', cliponaxis=False)
fig.update_yaxes(showticklabels=False)
fig.update_layout(legend = dict(
                        orientation='h',
                        title='',
                        font = dict(
                        family="Courier",
                        size=12,
                        ),
                        bgcolor='olive',
                        bordercolor='blue',
                        borderwidth=.5
                    ))

>**Notes**: 
>
> Back in 2018, Nigeria made the highest sales ($19.5M) amongst all country, but in 2020 it was the country with least sales (11.9M)
>
> Let's look at the profit for more insights 

### Profit by Country

#### Query the database

In [30]:
query = '''
SELECT 
    country, 
    year,
    SUM(profit) profit
FROM 
    breweries_sales
GROUP BY 
    country,
    year
ORDER BY 
    year DESC, profit DESC
'''
cntry_profit = pd.read_sql_query(query, engine)
cntry_profit.head()

country  year     profit
0    Ghana  2020  7144070.0
1  Senegal  2020  6687560.0
2     Togo  2020  6109960.0
3    Benin  2020  5273340.0
4  Nigeria  2020  4805320.0

#### Visualize the result

In [31]:
fig = px.bar(cntry_profit, 'country', 'profit', text_auto='$.3s', facet_row='year',
             title='Profit by Country', height=700, width=600, labels={'profit':''})
fig.update_traces(textposition='outside', cliponaxis=False)
fig.update_yaxes(showticklabels=False)

>**Notes**: 
>
> We saw earlier that Nigeria recorded the least sale ($11.9M) in year 2020, which happens to be the least sales across all years. This resulted to having the lowest profit ($4.81M) in year 2020, which is the least profit across all years. Nigeria produced the highest profit in 2018 and 2019. For the latest year, 2020, Ghana generated the highest profit ($7.14M) 

### Creating tables for different Countries

In [32]:
%%sql
CREATE TABLE nigeria AS
    SELECT  *
    FROM    breweries_sales
    WHERE   Country = 'Nigeria';

CREATE TABLE senegal AS
    SELECT  *
    FROM    breweries_sales
    WHERE   Country = 'Senegal';
    
CREATE TABLE benin AS
    SELECT  *
    FROM    breweries_sales
    WHERE   Country = 'Benin';

CREATE TABLE ghana AS
    SELECT  *
    FROM    breweries_sales
    WHERE   Country = 'Ghana';

CREATE TABLE togo AS
    SELECT  *
    FROM    breweries_sales
    WHERE   Country = 'Togo';

 * postgresql://postgres:***@localhost/international_breweries
(psycopg2.errors.DuplicateTable) relation "nigeria" already exists

[SQL: CREATE TABLE nigeria AS
    SELECT  *
    FROM    breweries_sales
    WHERE   Country = 'Nigeria';]
(Background on this error at: https://sqlalche.me/e/14/f405)


### Quantity Sold by Sales Rep in each Country.

#### Query the database

In [33]:
query = '''
WITH n AS (
    SELECT
        sales_rep,
        SUM(profit) t_prft
    FROM
        nigeria
    GROUP BY 
        sales_rep
),

s AS (
    SELECT
        sales_rep,
        SUM(profit) t_prft
    FROM
        senegal
    GROUP BY 
        sales_rep
),

b AS (
    SELECT
        sales_rep,
        SUM(profit) t_prft
    FROM
        benin
    GROUP BY 
        sales_rep
),

g AS (
    SELECT
        sales_rep,
        SUM(profit) t_prft
    FROM
        ghana
    GROUP BY 
        sales_rep
),

t AS (
    SELECT
        sales_rep,
        SUM(profit) t_prft
    FROM
        togo
    GROUP BY 
        sales_rep
)

SELECT 
    n.sales_rep, n.t_prft n_prft, s.t_prft s_prft, b.t_prft b_prft, g.t_prft g_prft, t.t_prft t_prft
FROM
    n
JOIN
    s USING(sales_rep)
JOIN
    b USING(sales_rep)
JOIN
    g USING(sales_rep)
JOIN
    t USING(sales_rep)
ORDER BY
    n_prft DESC
'''

prft = pd.read_sql_query(query, engine)
prft

sales_rep     n_prft     s_prft     b_prft     g_prft     t_prft
0      Jones  3488430.0  4638230.0  3516660.0  4518290.0  2609220.0
1       Gill  2769250.0  2795870.0  1964470.0  2877880.0  1882660.0
2    Jardine  2633290.0  2811950.0  2769560.0  2724190.0  2592660.0
3    Sorvino  2177510.0  1913770.0  2707290.0  1297600.0  2743530.0
4    Andrews  2075960.0  1944720.0  2606840.0  1572160.0  2383090.0
5     Morgan  1833820.0   977800.0  1599300.0  1590660.0  1956500.0
6      Smith  1673200.0  1068040.0  1151860.0  1414160.0  1682910.0
7     Parent  1406140.0  1137210.0  1269190.0  1257560.0  1832610.0
8     Kivell  1229670.0  1631650.0  1413280.0  1255810.0  1252290.0
9   Thompson  1047880.0  1159000.0  1100630.0  1425090.0  1318730.0
10    Howard  1041060.0  1406950.0   862890.0  1079650.0   496800.0

#### Visualize the result

In [34]:
fig = go.Figure()

fig.add_trace(
    go.Scatter(x=prft.sales_rep, y=prft.n_prft, name='Nigeria')
)
fig.add_trace(
    go.Scatter(x=prft.sales_rep, y=prft.s_prft, name='Senegal')
)
fig.add_trace(
    go.Scatter(x=prft.sales_rep, y=prft.b_prft, name='Benin')
)
fig.add_trace(
    go.Scatter(x=prft.sales_rep, y=prft.g_prft, name='Ghana')
)
fig.add_trace(
    go.Scatter(x=prft.sales_rep, y=prft.t_prft, name='Togo')
)
fig.update_layout(width=900, title= '<b>Profit by Sales Rep in each Country<b>')

>**Notes**: 
>
> Howard made the lowest profit across all countries, except in senegal, where the profit is higher than that of Thompson, Parent, Smith, and Morgan. 

### Quantity and Profit made by Sales Rep in Nigeria.

#### Query the database

In [35]:
query = '''
SELECT
    sales_rep,
    SUM(quantity_sold) t_qty,
    SUM(total_cost) t_cost,
    SUM(total_sale) t_sale,
    SUM(profit) t_profit,
    ROUND(AVG(roi), 2) avg_roi
FROM
    nigeria
GROUP BY
    sales_rep
ORDER BY 
    t_profit DESC
'''

ng_sl = pd.read_sql_query(query, engine)
ng_sl

sales_rep  t_qty     t_cost     t_sale   t_profit  avg_roi
0      Jones  33104  5631170.0  9119600.0  3488430.0    56.88
1       Gill  21025  3502700.0  6271950.0  2769250.0    74.88
2    Jardine  21958  3447760.0  6081050.0  2633290.0    72.69
3    Sorvino  17998  2495590.0  4673100.0  2177510.0    82.85
4    Andrews  18337  2612390.0  4688350.0  2075960.0    75.03
5     Morgan  13893  1941530.0  3775350.0  1833820.0    89.03
6      Smith  12101  1821450.0  3494650.0  1673200.0    86.48
7     Parent  10874  1485510.0  2891650.0  1406140.0    89.29
8     Kivell  11875  1745080.0  2974750.0  1229670.0    68.71
9   Thompson   9092  1281270.0  2329150.0  1047880.0    77.76
10    Howard   9050  1578140.0  2619200.0  1041060.0    60.12

#### Visualize the result

In [36]:
from plotly.subplots import make_subplots

In [37]:
# Create figure with secondary y-axis
fig = make_subplots(specs=[[{"secondary_y": True}]])

fig.add_trace(
    go.Bar(x=ng_sl.sales_rep, y=ng_sl.t_profit, name='Profit'), secondary_y=False
)
fig.add_trace(
    go.Scatter(x=ng_sl.sales_rep, y=ng_sl.t_qty, name='Quantity'), secondary_y=True
)
fig.update_layout(width=900, title_text="<b>Quantity and Profit by Sales Rep in Nigeria<b>")
fig.update_yaxes(title_text="<b>Profit</b>", secondary_y=False)
fig.update_yaxes(title_text="<b>Quantity</b>", secondary_y=True)
fig.show()

>**Notes**: 
>
> In Nigeria, Jones sold the highest quantity followed by Gill. Though Gill made the second highest quantity sold, Jardine Profit was higher than Gill. 

### Quantity and Profit made by Sales Rep in Senegal.

#### Query the database

In [38]:
query = '''
SELECT
    sales_rep,
    SUM(quantity_sold) t_qty,
    SUM(total_cost) t_cost,
    SUM(total_sale) t_sale,
    SUM(profit) t_profit,
    ROUND(AVG(roi), 2) avg_roi
FROM
    senegal
GROUP BY
    sales_rep
ORDER BY
    t_profit DESC
'''

se_sl = pd.read_sql_query(query, engine)
se_sl

sales_rep  t_qty     t_cost     t_sale   t_profit  avg_roi
0      Jones  30660  4529170.0  9167400.0  4638230.0    96.83
1    Jardine  23416  3521750.0  6333700.0  2811950.0    77.34
2       Gill  20529  2914480.0  5710350.0  2795870.0    89.29
3    Andrews  18751  2985030.0  4929750.0  1944720.0    61.53
4    Sorvino  19429  3253780.0  5167550.0  1913770.0    54.66
5     Kivell  12036  1982500.0  3614150.0  1631650.0    75.39
6     Howard   8342  1191250.0  2598200.0  1406950.0   107.50
7   Thompson  10011  1498650.0  2657650.0  1159000.0    73.57
8     Parent  12354  1886790.0  3024000.0  1137210.0    57.79
9      Smith  12143  1900210.0  2968250.0  1068040.0    53.92
10    Morgan  13257  2043100.0  3020900.0   977800.0    47.27

In [39]:
# Create figure with secondary y-axis
fig = make_subplots(specs=[[{"secondary_y": True}]])

fig.add_trace(
    go.Bar(x=se_sl.sales_rep, y=se_sl.t_profit, name='Profit'), secondary_y=False
)
fig.add_trace(
    go.Scatter(x=se_sl.sales_rep, y=se_sl.t_qty, name='Quantity'), secondary_y=True
)
fig.update_layout(width=900, title_text="<b>Quantity and Profit by Sales Rep in Senengal<b>")
fig.update_yaxes(title_text="<b>Profit</b>", secondary_y=False)
fig.update_yaxes(title_text="<b>Quantity</b>", secondary_y=True)
fig.show()